Add support for EPD 3in7

mangelajo opened the pull request:
Matches implementation from waveshare/e-Paper code

https://github.com/waveshare/e-Paper/blob/master/RaspberryPi_JetsonNano/c/lib/e-Paper/EPD_3in7.c

Co-Authored-By: Miguel Angel Ajo <miguelangel@ajo.es>

mangelajo commented:
ok, it's ignoring #![allow(non_camel_case_types)]

I guess I will have to change the enum names.

caemor commented on src/epd3in7/command.rs:
```diff
@@ -0,0 +1,89 @@
+//! SPI Commands for the Waveshare 3.7" E-Ink Display
+#![allow(non_camel_case_types)]
```
Please remove the ignore non camel case types and switch the command names to CamelCase type similar as elsewhere in the repository (e.g. https://github.com/caemor/epd-waveshare/blob/main/src/epd2in13bc/command.rs)

caemor replied:
I should have read your comment first :see_no_evil: :-D

mangelajo replied:
Ohh, no worries, done :-)

mangelajo commented:
@caemor is there any driver doing grayscale or shall I define a new color type?

mangelajo commented:
Ah, there is Gray4! https://docs.rs/embedded-graphics/latest/embedded_graphics/pixelcolor/struct.Gray4.html

peckpeck commented:
I'd suggest to make the Color a generic parameter and vary it in impl, this way you could factor some of he code

mangelajo commented:
> I'd suggest to make the Color a generic parameter and vary it in impl, this way you could factor some of he code

Yes, I was thinking about that, but there are a couple of things holding me back:

* The code paths and initialization for both modes (see Gray1/Gray4 functions here [1]) are wildly different, they only seem to share SPI bus access.
* It would special-case this driver in terms of instantiation, I guess we could may be have a default instance in one of the modes -not sure if it's possible, new to rust here-, for other drivers (tri-color vs bw, the drivers are duplicated)

[1] https://github.com/waveshare/e-Paper/blob/master/RaspberryPi_JetsonNano/c/lib/e-Paper/EPD_3in7.c#L171

mangelajo commented:
Oh, I will go and implement partial updates for this driver first, while I keep rumiating about the Grayscales ;)

peckpeck commented:
If you use a generic type for color, you can easily separate the code for init with an impl for `Color` and and impl for `Gray4`. But update_frame will still be difficult, maybe smaller trait would help implementing this kind of cases.

peckpeck commented:
If there is no real improvement besides memory consumption, why have a separate BW initialization ? A BW update_frame on a gray initialized device should work no ?

mangelajo commented:
> If there is no real improvement besides memory consumption, why have a separate BW initialization ? A BW update_frame on a gray initialized device should work no ?

I really have no numbers to compare yet. I will eventually be working on Gray4 support on a different PR. At that point we can analyze it from both the timing and memory perspective and decide.

But it sounds like a good point, if they were to have the same refresh rates, why keep two code paths :)

mangelajo commented:
Note: I may need some time to get into Gray4, currently I am working on the application that uses the implementation of this PR.

Hopefully we can merge this if everything looks good, and then treat Gray4 support as an upgrade PR.

peckpeck commented:
Indeed, since we in 0.x releases we should merge things quickly to allow for easier testing.

peckpeck commented on src/epd3in7/mod.rs:
```diff
@@ -0,0 +1,258 @@
+/// Display with Fullsize buffer for use with the 2in9 EPD
+#[cfg(feature = "graphics")]
+pub type Display3in7 =
+    crate::graphics::Display<WIDTH, HEIGHT, false, { WIDTH as usize * HEIGHT as usize / 8 }, Color>;
```
As done in https://github.com/caemor/epd-waveshare/pull/128 you should prefer using buffer_len here
` { buffer_len(WIDTH as usize, HEIGHT as usize) }`

caemor replied:
```suggestion
    crate::graphics::Display<WIDTH, HEIGHT, false,  buffer_len(WIDTH as usize, HEIGHT as usize), Color>;
```

caemor commented on src/epd3in7/mod.rs:
```diff
@@ -0,0 +1,258 @@
+
+/// Display with Fullsize buffer for use with the 2in9 EPD
+#[cfg(feature = "graphics")]
+pub type Display3in7 =
+    crate::graphics::Display<WIDTH, HEIGHT, false, buffer_len(WIDTH as usize, HEIGHT as usize), Color>;
```
```suggestion
pub type Display3in7 = crate::graphics::Display<
    WIDTH,
    HEIGHT,
    false,
    { buffer_len(WIDTH as usize, HEIGHT as usize) },
    Color,
>;
```

mangelajo commented on src/epd3in7/mod.rs:
```diff
@@ -0,0 +1,263 @@
+
+const IS_BUSY_LOW: bool = false;
+
+/// Display with Fullsize buffer for use with the 2in9 EPD
```
2in9->3in7

It doesn't let me make code suggestions .

peckpeck replied:
isn't this your code ?

mangelajo replied:
lol, probably, I thought it was introduced in the changes afterwards.

caemor replied:
I was able to make the code suggestion that you wanted.

mangelajo commented:
> > I am not sure what's going on, I see the errors here:
> 
> You need to rebase your code, the main branch has been updated with new commits that impact your code. Something like `git rebase -i origin/main`

Ack, I will rebase and let's see how it goes, I would have expected the unit testing/etc to pass even if it needed rebase, of course github may complain if it's unable to merge your changes.

caemor commented on src/epd3in7/mod.rs:
```diff
@@ -0,0 +1,263 @@
+
+const IS_BUSY_LOW: bool = false;
+
+/// Display with Fullsize buffer for use with the 2in9 EPD
```
```suggestion
/// Display with Fullsize buffer for use with the 3in7 EPD
```

mangelajo commented:
Ok, I rebased and squashed the commits, now at least I get it reproduced locally:

```
error[E0407]: method `is_busy` is not a member of trait `WaveshareDisplay`
   --> src/epd3in7/mod.rs:260:5
    |
260 | /     fn is_busy(&self) -> bool {
261 | |         self.interface.is_busy(IS_BUSY_LOW)
262 | |     }
    | |_____^ not a member of trait `WaveshareDisplay`

error[E0277]: the trait bound `DELAY: _embedded_hal_blocking_delay_DelayUs<u32>` is not satisfied
  --> src/epd3in7/mod.rs:50:37
   |
50 | impl<SPI, CS, BUSY, DC, RST, DELAY> InternalWiAdditions<SPI, CS, BUSY, DC, RST, DELAY>
   |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `_embedded_hal_blocking_delay_DelayUs<u32>` is not implemented for `DELAY`
   |
note: required by a bound in `traits::InternalWiAdditions`
  --> src/traits.rs:36:12
   |
29 | pub(crate) trait InternalWiAdditions<SPI, CS, BUSY, DC, RST, DELAY>
   |                  ------------------- required by a bound in this
...
36 |     DELAY: DelayUs<u32>,
   |            ^^^^^^^^^^^^ required by this bound in `traits::InternalWiAdditions`
help: consider further restricting this bound
   |
58 |     DELAY: DelayMs<u8> + embedded_hal::prelude::_embedded_hal_blocking_delay_DelayUs<u32>,
   |                        ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

error[E0277]: the trait bound `DELAY: _embedded_hal_blocking_delay_DelayUs<u32>` is not satisfied
   --> src/epd3in7/mod.rs:125:37
    |
125 | impl<SPI, CS, BUSY, DC, RST, DELAY> WaveshareDisplay<SPI, CS, BUSY, DC, RST, DELAY>
    |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `_embedded_hal_blocking_delay_DelayUs<u32>` is not implemented for `DELAY`
    |
note: required by a bound in `traits::WaveshareDisplay`
   --> src/traits.rs:146:12
    |
139 | pub trait WaveshareDisplay<SPI, CS, BUSY, DC, RST, DELAY>
    |           ---------------- required by a bound in this
...
146 |     DELAY: DelayUs<u32>,
    |            ^^^^^^^^^^^^ required by this bound in `traits::WaveshareDisplay`
help: consider further restricting this bound
    |
133 |     DELAY: DelayMs<u8> + embedded_hal::prelude::_embedded_hal_blocking_delay_DelayUs<u32>,
    |                        ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
```

I will try to have it fixed by tomorrow.

caemor commented:
If it helps:
#125 and #131 should be the issues:
- Replace DelayMs<u8> with DelayUs<u32> 
- And replace is_busy with wait_until_idle in the Waveshare Trait Implementation: ![image](https://user-images.githubusercontent.com/11088935/206298426-4e76ed07-19e5-4985-9b30-0d5f7b127a1e.png)

mangelajo commented:
> If it helps: #125 and #131 should be the issues:
> 
>     * Replace DelayMs with DelayUs
> 
>     * And replace is_busy with wait_until_idle in the Waveshare Trait Implementation: ![image](https://user-images.githubusercontent.com/11088935/206298426-4e76ed07-19e5-4985-9b30-0d5f7b127a1e.png)

Thanks a lot, it helps, yes! #59

mangelajo commented:
hopefully compile now after rebase :-)

mangelajo commented:
> Thanks for your great work!

Thank you!! 🙏🏻
